### PR TITLE
tests: more testing for BigTable

### PR DESCRIPTION
**Pull request by justinsb**
- **tests: record more accurate bigtable golden log**
- **tests: create scenario tests for bigtableinstance**
- **tests: run all scenario tests in tests-e2e-scenarios**
- **tests: support MockGCP-backdoor command**
- **mockgcp: improve fidelity of BigTable**
- **tests: add support for "TOUCH" magic**

### CODE REVIEW

File: pkg/test/resourcefixture/testdata/basic/bigtable/v1beta1/bigtableinstance/bigtableinstanceautoscaling2/_http.log [hunk 1]
```diff
@@ -103,66 +103,18 @@ OK
     "originalRequest": {
       "clusters": {
         "us-central1-a": {
-          "clusterConfig": {
```
Comment (cheftako) [inline pkg/test/resourcefixture/testdata/basic/bigtable/v1beta1/bigtableinstance/bigtableinstanceautoscaling2/_http.log:106]: I like what we're adding but a little concerned that this means we are losing coverage on our tests for a few field types.

**Comment by cheftako** [comment]
2024-07-27T01:48:06.8934676Z     utils.go:99: unexpected diff in /home/runner/work/k8s-config-connector/k8s-config-connector/pkg/test/resourcefixture/testdata/basic/bigtable/v1beta1/bigtableinstance/bigtableinstanceautoscaling2/_http.log:   strings.Join({
2024-07-27T01:48:06.8936875Z           	... // 3433 identical bytes
2024-07-27T01:48:06.8937872Z           	": \"2024-04-01T12:34:56.123456Z\"\n  },\n  \"name\": \"operations/proje",
2024-07-27T01:48:06.8938993Z           	"cts/${projectId}/instances/bt-${uniqueId}/locations/us-central1-",
2024-07-27T01:48:06.8939812Z         - 	"a",
2024-07-27T01:48:06.8942160Z         + 	"c",
2024-07-27T01:48:06.8943585Z           	"/operations/${operationID}\"\n}\n\n---\n\nGRPC /google.longrunning.Ope",
2024-07-27T01:48:06.8945209Z           	"rations/GetOperation\n\n{\n  \"name\": \"operations/projects/${project",
2024-07-27T01:48:06.8946545Z           	"Id}/instances/bt-${uniqueId}/locations/us-central1-",
2024-07-27T01:48:06.8947354Z         - 	"a",
2024-07-27T01:48:06.8947866Z         + 	"c",
2024-07-27T01:48:06.8948932Z           	"/operations/${operationID}\"\n}\n\nOK\n\n{\n  \"done\": true,\n  \"metadata",
2024-07-27T01:48:06.8950342Z           	"\": {\n    \"@type\": \"type.googleapis.com/google.bigtable.admin.v2.",
2024-07-27T01:48:06.8951302Z           	... // 1004 identical bytes
2024-07-27T01:48:06.8952433Z           	": \"2024-04-01T12:34:56.123456Z\"\n  },\n  \"name\": \"operations/proje",
2024-07-27T01:48:06.8953742Z           	"cts/${projectId}/instances/bt-${uniqueId}/locations/us-central1-",
2024-07-27T01:48:06.8954543Z         - 	"a",
2024-07-27T01:48:06.8954986Z         + 	"c",
2024-07-27T01:48:06.8955973Z           	"/operations/${operationID}\",\n  \"response\": {\n    \"@type\": \"type.",
2024-07-27T01:48:06.8957860Z           	"googleapis.com/google.bigtable.admin.v2.Instance\",\n    \"displayN",
2024-07-27T01:48:06.8958876Z           	... // 34901 identical bytes
2024-07-27T01:48:06.8959492Z           }, "")

**Comment by justinsb** [comment]
Yes, it looks like maybe our existing controllers are non-deterministic :-(

**Comment by justinsb** [comment]
> Yes, it looks like maybe our existing controllers are non-deterministic :-(

Actually, I think I found the (main?) non-determinism ... I was picking a zone for the operation based on the map ordering, which is not stable.  (The fact that the operation is zonal is sort of weird, but ... it's what GCP does...)

File: mockgcp/mock_http_roundtrip.go [hunk 2]
```diff
@@ -94,7 +96,23 @@ type MockService interface {
...
+	// NewGRPCConnection returns a grpc connection to our mock implementation
+	NewGRPCConnection(ctx context.Context) *grpc.ClientConn
+
+	// We can dispatch test commands
+	SupportsTestCommands
+}
+
+type SupportsTestCommands interface {
+	// RunTestCommand is a "backdoor" into our mock implementation that is useful for fault injection or faking scaling events etc
+	RunTestCommand(ctx context.Context, service string, command string) error
```
Comment (yuwenma) [inline mockgcp/mock_http_roundtrip.go:112]: Suggest calling out that the `RunTestCommand` pairs with 'MockGCPBackdoor' env var (unless we really want to keep this feature silent 🤫 )

File: mockgcp/mockbigtable/cluster.go [hunk 3]
```diff
@@ -219,7 +230,13 @@ func (s *MockService) populateDefaultsForCluster(obj *pb.Cluster) error {
...
 	if obj.ServeNodes == 0 {
-		obj.ServeNodes = 2
+		autoscaling := obj.GetClusterConfig().ClusterAutoscalingConfig
+		if autoscaling != nil {
+			obj.ServeNodes = autoscaling.AutoscalingLimits.GetMinServeNodes()
+		}
+		if obj.ServeNodes == 0 {
+			obj.ServeNodes = 2
+		}
 	}
```
[the remark this answers is not in the document]
Comment (justinsb) [reply mockgcp/mockbigtable/cluster.go:240]: Yes, BigTable seems to default to 2 (AFAICT) without Autoscaling.  We just want to keep record-gcp and compare-mock as close as possible

File: pkg/test/resourcefixture/testdata/basic/bigtable/v1beta1/bigtableinstance/bigtableinstanceautoscaling2/_generated_object_bigtableinstanceautoscaling2.golden.yaml [hunk 4]
```diff
@@ -22,7 +22,7 @@ spec:
       minNodes: 3
       storageTarget: 2560
     clusterId: us-central1-a
-    numNodes: 2
+    numNodes: 3
```
[the remark this answers is not in the document]
Comment (justinsb) [reply pkg/test/resourcefixture/testdata/basic/bigtable/v1beta1/bigtableinstance/bigtableinstanceautoscaling2/_generated_object_bigtableinstanceautoscaling2.golden.yaml:25]: So numNodes is the KCC field that is called serveNodes in the GCP API.  We bumped it with our backdoor call.

File: mockgcp/mock_http_roundtrip.go (hunk 2 above)
Comment (justinsb) [reply mockgcp/mock_http_roundtrip.go:112]: Good call, I'll add it (in a follow-on PR)

**Pull request merged**